asdf/extension/_converter.py

```diff
@@ -211,9 +212,25 @@ def __init__(self, delegate, extension):
 
         self._types = []
 
-        if not len(self._tags) and not hasattr(delegate, "select_tag"):
-            # this converter supports no tags so don't inspect the types
-            return
+        if not len(self._tags):
+            # this converter supports no tags supported by the extension
+            # before indexing it's types we need to check select_tag
+
+            # check if it implements select_tag (and not just because it
+            # subclasses Converter)
+            if not hasattr(delegate, "select_tag"):
+                return
+
+            # find which class implements select_tag
+            for class_ in inspect.getmro(delegate.__class__):
+                if "select_tag" in class_.__dict__:
+                    if class_ is not Converter:
+                        # a non-Converter class implements select_tag so consider the types for this Converter
+                        break
+                    else:
+                        # this converter implements select_tag only because it inherits from Converter
+                        # don't index it's types
+                        return
```

Comment:
I think it should work to add an exception to the default `Converter.select_tag` method so that classes that inherit from `Converter` but are called with no tags have a more informative error (it was a puzzling one to untangle).

As far as supporting converters in extensions which support tags which aren't in the extension/manifest, it seems like a good thing to drop (although we should likely wait until stdatamodels is released which seems very likely to happen before asdf 3.0).

Dropping the gracious handling of the stdatamodels bug has the big benefit of allowing us to remove the more complicated part of this PR (the code using `inspect.getmro`). However, I do like that this catches the error (a tagless converter that doesn't re-implement `select_tag`) when the converter is proxied instead of when the converter gets used.

One thought that comes to mind would be to remove `select_tag` from the abstract `Converter`. This has one downside of removing the easy documentation of `select_tag` (is there another way we could document that method, ideally in a way that is runtime inspectable for someone testing out writing a converter in the repl?).

Let me make an attempt at updating this PR to:
- add an exception when the default `select_tag` is called with no tags
- drop the code supporting the ignoring of converters that support tags not mentioned in the manifest

@zacharyburnett do you know if there is a plan to release stdatamodels any time soon? The tldr; is that it relied on what is likely a bug in asdf which this PR fixes. The main branch of stdatamodels was fixed in https://github.com/spacetelescope/stdatamodels/pull/184 which has not yet made it into a release.

Reply:
It looks like removing the default `Converter.select_tag` did the trick.

I moved and slightly adapted the `select_tag` documentation to the `Converter` documentation (see: https://github.com/asdf-format/asdf/pull/1594/commits/18a3cc77b16d9ffa3c834054f38bc90bfcaa2edd#diff-14b06e3da444949d9a2c7f7b737b99b6c4112ede41f041a6d4f57b3627574d67R17).
I also changed the links to `select_tag` in the docs to link to `Converter` as that docstring now contains extra information about what is required of a `select_tag` method.

See the `Converter` docs here: https://asdf--1594.org.readthedocs.build/en/1594/api/asdf.extension.Converter.html#asdf.extension.Converter
and one of the modified links near the end of the section here: https://asdf--1594.org.readthedocs.build/en/1594/asdf/extending/converters.html#the-converter-interface

The devdeps failures appear unrelated and caused numpy 2.0 deprecating `maximum_sctype` https://github.com/numpy/numpy/pull/24154 which is used in astropy: https://github.com/astropy/astropy/blob/cd9dc7f818d97a64c2b5657a93fa3740a2270a38/astropy/nddata/bitmask.py#L22

Reply:
I re-added the above check (which causes the gwcs step and wcs converters to fail this check). I think the manifest in asdf-transform-schemas should be split into 2 versions and gwcs updated to register 2 extensions. I will put this PR into draft until those changes can be made because without them this causes many test failures due to warnings getting turned into errors.